[GH-6366] Add functionality to disable Authentication when connecting to SMTP

moonmeister opened the pull request:
Issue #6366: add disable auth to SMTP settings

Not sure if I need a "Help Wanted" ticket for this, but I think the work is done. 

The "None" option under "Connection Security" now seems redundant. Should I remove that option as part of this PR? Thanks.

#### Summary
Add a boolean Choice box to disable Authentication for the SMTP server.

#### Ticket Link
https://github.com/mattermost/platform/issues/6366

#### Checklist
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates

jasonblais commented:
Thanks @moonmeister,

Do the other services you use disable authentication as well for SMTP?
Why is the "None" option under "Connection Security" now redundant?

moonmeister commented:
1. I relay (or have relayed) printers, gitlab, graylog, firewalls and probably a few other things through google's SMTP relay and I've never had this issue. They usually ask whether authentication is needed.

2. The none option is redundant based on the it's functionality. It uses an unencrypted or "Insecure" connection just like the "PLAIN" option except it skips the authentication step. Now that we have a dedicated option to disable authentication the "None" option should probably be remove.

jasonblais commented:
Thanks @moonmeister for the reply, really helpful.

1. Is it correct to say that previously Mattermost supported unencrypted, unauthenticated SMTP server setup (via "None" option), and this pull request would add support for encrypted, but unauthenticated SMTP setup?

2. If we removed the "None" option, what would happen to existing installs that use the option? I worry their SMTP setup would break.

moonmeister commented:
@jasonblais My pleasure!

1. Yes, that would be correct.

2. Without modifying the code and finding out my guess is nothing. The code in mail.go uses "if" statements to do the authentication and negotiate TLS if those options are set and does nothing if none is set. My suspicion is that it won't break anything. I can work on making the change though and test the hypothesis.

moonmeister commented:
Oops, Didn't realize that code change would auto update this PR (I'm new at this). Well i can revert the change if we don't want to remove the "None" option. Nothing seemed to break on the test I did. Please double check though.

moonmeister commented:
Okay I also simplified the logic around auth, should be faster and easier to read. Just let me know if these are okay changes and I can squash the commits. Thanks

moonmeister commented:
Looks like the "TestSendPasswordReset" test is broken. I'll take a look to see what would be causing this. Any pointers would be appreciated. I'm pretty busy the next couple days so it might be the weekend before I can get to resolving the issue. Thanks.

jasonblais commented:
Cool, thanks @moonmeister. Weekend works well, no rush. If you get stuck resolving the broken test, you're welcome to join our [Developers channel](https://pre-release.mattermost.com/core/channels/developers) in our contributors Mattermost instance -- there community members and the core team can help answer questions.

Just to clarify..if I have the ConnectionSecurity option set to `None`, then upgrade Mattermost, what will this option change to if the `None` option is no longer available?

moonmeister commented:
Okay, @jasonblais , I did some thinking, there is an issue with the change in configuration file. 

1. Assuming people aren't merging the new "default" config file into their existing config, their config won't define `config.EmailSettings.EnableSMTPAuth`. I believe this means the value becomes false,  then If they are using encrypted options they won't auth. I'm not sure how we could de

2. If they merge the new config, where the default is `true` and they are using `None` then they will Auth when they shouldn't. This could be fixed by 

For option 2 we could user this code:
`if config.EmailSettings.ConnectionSecurity == model.CONN_SECURITY_NONE {
		config.EmailSettings.ConnectionSecurity = model.CONN_SECURITY_PLAIN
		config.EmailSettings.EnableSMTPAuth = false
	}`

I have no idea how we'd detect option 1.

I'm thinking the best option would be a way to upgrade the config file when they upgrade their version, is this possible? If so where is the best place for that to happen in code?

moonmeister commented:
Well, the `TestSendPasswordReset` and `TestResetPassword` are now passing on my local computer after the change I made, not sure why they wouldn't be testing on the CI server. Any suggestions? Thanks.

moonmeister commented:
So, found the issue. When the CI server runs the config tests it over writes the "Default" test settings that I'd hard coded (in apitestlib.go) with what's in the the config file. I had incorrectly set the default value of `EnableSMTPAuth` to be true for the default test SMTP server. So, it was breaking. When I was doing one off tests it was fine because the config tests hadn't run. 

This reinforced the fact that the upgrade path was broken. To fix this I changed the the `EnableSMTPAuth` variable to be a pointer to a bool. This meant, when it was read in, if it wasn't in the config file, instead of getting the default value of `false` it was now `nil`. If I detected `nil` and `ConnectionSecurity` was `None`, I knew I could safely set `EnableSMTPAuth` to false and `ConnectionSecurity` to `PLAIN`. Otherwise if `ConnectionSecurity` was `TLS` or `STARTTLS` then `EnableSMTPAuth` got set to true. This will function fine and if they ever "save" their config the new settings will get saved and the logic will no longer run when they read in the config.json.

I've removed my settings changes to "apitestlib.go" and correctly setup the config file.

Builds are now passing! I think I've covered everything. Any other feedback?

jasonblais commented:
Thanks @moonmeister! Let me start a conversation in the [Developers channel](https://pre-release.mattermost.com/core/channels/developers) to get more thoughts on this PR first before we move on.

jasonblais commented:
Okay, so as [briefly discussed](https://pre-release.mattermost.com/core/pl/58h6w1e4o3rtfj4h6i6bhngb7o), the approach here is good.

Your comment sounds good to me as well. Thank you for taking steps to ensure the settings won't break on upgrade:
> This reinforced the fact that the upgrade path was broken. To fix this I changed the EnableSMTPAuth variable to be a pointer to a bool. This meant, when it was read in, if it wasn't in the config file, instead of getting the default value of false it was now nil. If I detected nil and ConnectionSecurity was NONE, I knew I could safely set EnableSMTPAuth to false and ConnectionSecurity to PLAIN. Otherwise if ConnectionSecurity was PLAIN, TLS, or STARTTLS then EnableSMTPAuth got set to true. This will function fine and if they ever "save" their config the new settings will get saved and the logic will no longer run when they read in the config.json.

The only change I'd recommend is keeping "NONE" and removing "PLAIN" instead. Someone looking at the options for the first time might not see the relationship between "Connection Security" and "Plain" but can pretty quickly do so for None, TLS and STARTTLS.

So in that case, the mapping would be
 - None --> Keep `ConnectionSecurity` as `None`, set `EnableSMTPAuth` to `False`
 - Plain --> Set `ConnectionSecurity` to `None`, set `EnableSMTPAuth` to `True`
 - TLS --> Keep `ConnectionSecurity` as `TLS `, set `EnableSMTPAuth` to `True`
 - STARTTLS --> Keep `ConnectionSecurity` as `STARTTLS `, set `EnableSMTPAuth` to `True`

with the new option of setting `EnableSMTPAuth` to `False` when encryption is enabled.

What do you think?

moonmeister commented:
@jasonblais Good idea, that makes it match the LDAP options too. So those two things are consistent.

moonmeister commented:
@jasonblais The "PLAIN" Option is nolonger in use by the front end or back end. The build error makes that clear.

`
/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/webapp/components/admin_console/connection_security_dropdown_setting.jsx
  29:7  error  'SECTION_PLAIN' is assigned a value but never used  no-unused-vars
`
Should I just comment out that code? Do we want to remove it all together? We could also do some code cleanup on the backend of the "PLAIN" Option.

jasonblais commented:
@crspeller @hmhealey Any thoughts?

> @jasonblais The "PLAIN" option is no longer in use by the frontend or backend. The build error makes that clear.

> /home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/webapp/components/admin_console/connection_security_dropdown_setting.jsx 29:7 error 'SECTION_PLAIN' is assigned a value but never used no-unused-vars
Should I just comment out that code? Do we want to remove it all together? We could also do some code cleanup on the backend of the "PLAIN" option.

jasonblais commented:
Checked with Chris -- lets remove the code since we're removing the option

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0bef0889a92169db3.spinmint.com:8065/pr6639

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0bef0889a92169db3

jasonblais commented:
Thanks @moonmeister for this! Asked for UI text review, and then it should be ready for the second stage (dev review of code + testing)

jasonblais commented on webapp/i18n/en.json:
```diff
@@ -298,6 +296,8 @@
   "admin.email.smtpUsernameExample": "E.g.: \"admin@yourcompany.com\", \"AKIADTOVBGERKLCBV\"",
   "admin.email.smtpUsernameTitle": "SMTP Server Username:",
+  "admin.email.enableSMTPAuthTitle": "Enable SMTP Authentication:",
+  "admin.email.enableSMTPAuthDesc": "When enabled the username and passsword below are used to authenticate to the SMTP server. Some servers require authentication to be disabled.",
```
Proposed help text:

```
When enabled, username and password are used for authenticating to the SMTP server.
```

Also, if you'd like, would you be able to change the help text for the following two settings? (I can also prepare a separate PR if you'd prefer, I just thought it's a good opportunity to improve the help texts, since the settings are related)

```
SMTP Server Username
Help text: The username for authenticating to the SMTP server.

SMTP Server Password
Help text: The password for the SMTP server username.
```

jasonblais commented on webapp/components/admin_console/email_settings.jsx:
```diff
@@ -201,6 +203,25 @@ export default class EmailSettings extends AdminSettings {
                     onChange={this.handleChange}
                     disabled={!this.state.sendEmailNotifications}
                 />
+                <BooleanSetting
```
Would you be okay moving the `Enable SMTP Authentication`, `SMTP Server Username` and `SMTP Server Password` settings below `SMTP Server Port`?

So the order of the settings would be 

```
SMTP Server
SMTP Server Port
Enable SMTP Authentication (true/false)
SMTP Server Username [disabled if auth = false]
SMTP Server Password [disabled if auth = false]
```

I think the flow is more intuitive since you first specify the SMTP server and port, then enable authentication (if you require it).

moonmeister commented:
@jasonblais I didn't really like my text anyway. I'm happy to make these changes, I think they all make good sense. The reordering is an excellent idea as well. I'll make these changes and get them pushed tomorrow hopefully. Thanks.

moonmeister commented:
@jasonblais FYI Changes have been completed. Also, if you don't mind my asking, what was the reason behind the milestone change?

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-087fd30376ecff753.spinmint.com:8065/pr6639

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-087fd30376ecff753

mattermod commented:
Spinmint test server destroyed

moonmeister commented:
@jasonblais Makes sense. Thanks.

hmhealey commented on model/config.go:
```diff
@@ -752,6 +753,18 @@ func (o *Config) SetDefaults() {
+		} else {
+			*o.EmailSettings.EnableSMTPAuth = true
+		}
+		if o.EmailSettings.ConnectionSecurity == CONN_SECURITY_PLAIN {
```
Can you move this block so that it's still ran when `EnableSMTPAuth` is not nil? I don't know if it's super necessary, but it would be nice to know that `ConnectionSecurity` can never equal `CONN_SECURITY_PLAIN` elsewhere

moonmeister replied:
I don't think I see an issue with doing this. In the rare case that the "EnableSMTPAuth" setting  has been implemented, but somehow this is still set to "Plain" this would catch that case as well.

hmhealey commented on utils/config.go:
```diff
@@ -434,6 +434,7 @@ func getClientConfig(c *model.Config) map[string]string {
 	props["EnableSignInWithUsername"] = strconv.FormatBool(*c.EmailSettings.EnableSignInWithUsername)
 	props["RequireEmailVerification"] = strconv.FormatBool(c.EmailSettings.RequireEmailVerification)
 	props["EnableEmailBatching"] = strconv.FormatBool(*c.EmailSettings.EnableEmailBatching)
+	props["EnableSMTPAuth"] = strconv.FormatBool(*c.EmailSettings.EnableSMTPAuth)
```
This shouldn't be in the client config since it only contains settings needed by the client outside of the System Console

moonmeister replied:
Okay, I'll remove it.

hmhealey commented on webapp/i18n/en.json:
```diff
@@ -286,7 +284,7 @@
   "admin.email.skipServerCertificateVerification.description": "When true, Mattermost will not verify the email server certificate.",
   "admin.email.skipServerCertificateVerification.title": "Skip Server Certificate Verification: ",
-  "admin.email.smtpPasswordDescription": " Obtain this credential from administrator setting up your email server.",
+  "admin.email.smtpPasswordDescription": "The password for the SMTP server username.",
```
@lfbrock I think this needs a text review

moonmeister replied:
@hmhealey @lfbrock I was asked previously by @jasonblais to make that text change. IT does read a bit odd though with the word "username" included. It could probably be dropped or maybe be changed to "associated with the username"

jasonblais replied:
The help text was already reviewed by lfbrock and jeffschering: https://pre-release.mattermost.com/core/pl/w6fqjd6dwirdznf54m4z61bbww

Have queued `The password associated with the SMTP username.` for review.

jasonblais replied:
@hmhealey @moonmeister Let's go with `The password associated with the SMTP username.` as the SMTP password help text

moonmeister replied:
@hmhealey @jasonblais Change done.

crspeller commented on model/config.go:
```diff
@@ -752,6 +753,19 @@ func (o *Config) SetDefaults() {
 		*o.EmailSettings.EmailBatchingInterval = EMAIL_BATCHING_INTERVAL
 	}
 
+	if o.EmailSettings.EnableSMTPAuth == nil {
```
This function is for setting defaults. I don't think this logic belongs here.

hmhealey replied:
It looks fine to me. It's setting a default, but it's doing it intelligently based on the other setting.

crspeller commented on utils/mail.go:
```diff
@@ -59,22 +60,17 @@ func newSMTPClient(conn net.Conn, config *model.Config) (*smtp.Client, *model.Ap
-			return nil, model.NewLocAppError("SendMail", "utils.mail.new_client.auth.app_error", nil, err.Error())
-		}
-	} else if config.EmailSettings.ConnectionSecurity == model.CONN_SECURITY_STARTTLS {
+	if config.EmailSettings.ConnectionSecurity == model.CONN_SECURITY_STARTTLS {
```
Do you think this might belong in connectToSMTPServer above?

moonmeister replied:
@crspeller I think it makes more sense to make sure the running configuration is compatible and correct as soon as possible. If we leave it to the `mail.go` file then when the front end (or any other existing or future reference) tries to pull these values they may or may not be correct depending on whether or not `mail.go` has executed since the server got turned up. 
So, for consistency I think the `config.go` is the best location to update outdated configurations options to make sure the configuration is correct across all aspects of the platform.

crspeller replied:
Ok. I agree. I think this comment was meant for my other feedback.

moonmeister replied:
@crspeller Oops, yeah evidently I miss read, The STARTTLS related code was already in this function. I simplified some of the logic but did not move/add the code.

moonmeister commented:
@hmhealey @crspeller @jasonblais @lfbrock What's next? Any other modifications? Thanks.

jasonblais commented:
I think @hmhealey and @crspeller just need to re-review and if the changes look good this PR should be ready to go

hmhealey commented on model/config.go:
```diff
@@ -752,6 +753,19 @@ func (o *Config) SetDefaults() {
+		}
+	}
+
+	if o.EmailSettings.ConnectionSecurity == CONN_SECURITY_PLAIN {
```
I think this one needs to come before the other block. If this is set to plain and it needs to configure EnableSMTPAuth, it'll do it incorrectly

moonmeister replied:
@hmhealey It definitely goes there. If you override the incorrect `ConnectionSecurity` setting from `None` to `Plain` then when it tries to set the `EnableSMTPAuth` it will set it incorrectly. `EnableSMTPAuth` must be set based on the user config before the config for `ConnectionSecurity` gets updated.

hmhealey replied:
Oh, I see. The previous behaviour worked as if the new `EnableSMTPAuth` setting was true when the `ConnectionSecurity` was plain. This makes sense then

moonmeister commented:
@crspeller Did you have anything else for me to change? thanks.

jasonblais commented:
@hmhealey By the way as discussed, this wasn't tested by a PM as it needs a dev to set up email servers.

Not sure if any testing was done on this PR?

Sorry @moonmeister that it has taken longer to review this PR than usual.
